src/huggingface_hub/inference_api.py

```diff
@@ -0,0 +1,145 @@
+    """
+
+    def __init__(
+        self,
+        repo_id: str,
+        task: Optional[str] = None,
+        token: Optional[str] = None,
+        gpu: Optional[bool] = False,
+    ):
+        """Inits headers and API call information.
+
+        Args:
+            repo_id (``str``): Id of repository (e.g. `user/bert-base-uncased`).
+            task (``str``, `optional`, defaults ``None``): Whether to force a task instead of using task specified in the repository.
+            token (:obj:`str`, `optional`):
+                The API token to use as HTTP bearer authorization. This is not the authentication token.
+                You can find the token in https://huggingface.co/settings/token. Alternatively, you can
+                find both your organizations and personal API tokens using `HfApi().whoami(token)`.
+            gpu (``bool``, `optional`, defaults ``False``): Whether to use GPU instead of CPU for inference(requires Startup plan at least).
+        .. note::
+            Setting :obj:`token` is required when you want to use a private model.
+        """
+        self.options = {"wait_for_model": True, "use_gpu": gpu}
+
+        self.headers = {}
+        if isinstance(token, str):
+            self.headers["Authorization"] = "Bearer {}".format(token)
+
+        # Configure task
+        model_info = HfApi(ENDPOINT_HUB).model_info(repo_id=repo_id, token=token)
+        if not model_info.pipeline_tag and not task:
+            raise ValueError(
+                "Task not specified in the repository. Please add it to the model card using pipeline_tag (https://huggingface.co/docs#how-is-a-models-type-of-inference-api-and-widget-determined)"
+            )
+
+        if task and task != model_info.pipeline_tag:
+            if task not in ALL_TASKS:
+                raise ValueError(f"Invalid task {task}. Make sure it's valid.")
+
+            logger.warning(
+                "You're using a different task than the one specified in the repository. Be sure to know what you're doing :)"
+            )
+            self.task = task
+        else:
+            self.task = model_info.pipeline_tag
+
+        self.api_url = f"{ENDPOINT}/pipeline/{self.task}/{repo_id}"
+        print(
+            f"Initialized Inference API for {repo_id} with task {self.task} for {model_info.library_name} library"
+        )
+
+    def __repr__(self):
+        items = (f"{k}='{v}'" for k, v in self.__dict__.items())
+        return f"{self.__class__.__name__}({', '.join(items)})"
+
+    def __call__(
+        self,
+        inputs: Union[str, Dict, List[str], List[List[str]]],
+        params: Optional[Dict] = None,
+    ):
+        payload = {
+            "inputs": inputs,
+            "options": self.options,
+        }
+
+        if params:
+            payload["parameters"] = params
+
+        # TODO: Decide if we should raise an error instead of
+        # returning the json.
```

Comment:
As mentioned offline with @osanseviero, I think handling exceptions here is the most pythonic way to handle things - and that without such logic there's little value in having a Python wrapper for the inference API if it's a glorified `cURL` utility. 

I would argue that the wrapper's ability to return appropriate errors such as `raise ModelNotLoadedError` or `raise MissingKeyError`, which inherit from the appropriate canonic errors (`ValueError`, `OSError`), is very important API-wise.